### DISCUSSION TITLE
Is there a way to display a tooltip?

### DISCUSSION THREAD

**Discussion by SebastienSyd** [Textualize/textual#2204] [Q&A] [answered]
The idea would be to display some text in a tooltip when hovering a Label. Is there a way to achieve this in Rich/Textual?

A concrete example is: we display a date in a human readable format (ex. 6 days ago) and when hovering the date label we get the full formatted date (ex. 2023-04-03 12:00:00Z)

Thank you

**Comment by TomJGooding** [comment]
I'm not sure this is currently possible with Textual - though I would love to be proven wrong!

For starters, unless I'm missing something I can't find any sort of `on_hover` handler in the API at the moment?

**Reply by TomJGooding** [reply]
After a bit more digging, I've found there are actually `on_mouse_move` and `on_leave` methods that seem to track whether the widget is currently hovered.

Perhaps if the tooltip text was created on a new layer and only displayed when the label is hovered...

**Reply by SebastienSyd** [reply]
thanks @TomJGooding for the idea, will play around with that and update this thread with news if that works out

**Reply by TomJGooding** [reply]
I've spent a bit of time hacking away at this myself but without much progress, so I would be really interested if you do find a solution!

Ideally I think you'd want to wrap everything nicely inside a `Tooltip` widget, but after experimenting both with a custom `Widget` and a `Label` subclass, there's some issues with layers and offset that I'm struggling to reconcile.

**Comment by 1j01** [comment, marked as answer]
FYI, [tooltips are now part of the library](https://textual.textualize.io/blog/2023/06/01/textual-adds-sparklines-selection-list-input-validation-and-tool-tips/)!

Just set [`widget.tooltip`](https://textual.textualize.io/guide/widgets/#tooltips) to a string or other renderable, and you can style the `Tooltip` widget it creates.
